### ISSUE TITLE
Nextcloud integration

### ISSUE THREAD

**Issue by mysticmode** [joyread/server#5]
Here is the process and my thoughts on this:

There is an admin and the users - So, Admin will register for Nextcloud integration in Joyread. And basically other users will come via invitation process. So if the admin invite a new user to Joyread, there will be an option where it will ask whether the user who is going to be invited should upload books locally or should have his/her own Nextcloud instance to sync ebooks. If the admin chooses Nextcloud - then the user must have to have her own Nextcloud instance in order to share ebooks in Joyread.

The main point of integrating Nextcloud is that the syncing feature. It is better than uploading. So, all the users other than admin should also be able to sync their own Nextcloud books.

Another thing is that - The source of the ebook will be kept wherever it is coming from. For eg: You are an admin and you are inviting me. Then I'm syncing my ebooks and sharing it with you. That ebooks will only come from my Nextcloud instance. In other words, it won't be present in your Nextcloud instance eventhough it is shared with you. So, when you are reading that ebook, it will be pulled from my instance. If you want to keep the books that are shared with you locally then you can click the download option in Joyread and do that.

Hope that makes sense. Please let me know if it is not clear or I'm missing something. Thanks!

**Comment by mysticmode** [comment]
Admin might have a limited storage solution in the box where Joyread is installed. So, I think we should keep this option to the admin.

**Comment by GAumala** [comment]
In that case I think it would be better to set a limit of disk space for each user. If you as an admin are already sending invites manually, You can avoid running out of disk space without too much trouble.

**Comment by mysticmode** [comment]
Exactly, that was my plan. Sorry I didn't mention this. So, the disk space limit option will be shown when the admin chooses the user to upload books locally in the instance. This is quite important as you said.

**Comment by mysticmode** [comment]
@GAumala I've added it as a new issue [here](https://github.com/joyread/server/issues/8)

Thanks for mentioning!

**Comment by mysticmode** [comment]
I tried to use PDF.js (also viewer.js) and added PDF file that is hosted on Nextcloud instance. But nextcloud doesn't allow CORS.

I tried checking in their forums, but no use. Let's see - I'll try to post to some developers regarding this.

If this doesn't work - then we have to put Nextcloud integration on hold until they added this feature.

**Comment by sunjam** [comment]
Perhaps @rullzer and @schiessle from Nextcloud for WebDAV or @jancborchardt can offer some insight

**Comment by jancborchardt** [comment]
@mysticmode that sounds cool, and integration would be great! :) If you need anything specific from the Nextcloud server, it’s best to open an issue in our repository at https://github.com/nextcloud/server/issues/ (that’s where developers hang out, the forum is mostly for user support).

Generally Nextcloud works simply using WebDAV, so using that for syncing and then combining it with the app passwords mechanism we offer is probably best.

But @rullzer @schiessle indeed have to give more pointers here, I’m not that deep in the code. :slightly_smiling_face:

**Comment by mysticmode** [comment]
@jancborchardt Thank you! I'll post it on the Nextcloud server repo issues with the detailed comment on what problem I'm facing in a few days. Also, I will post the same here as well. Thanks again

**Comment by mysticmode** [comment]
I have integrated Nextcloud successfully yesterday. But there are few things we need to consider.

Fetching the books from Nextcloud and directly reading it will be a very slow process and consider that I'm using PDF.js for PDFs which itself takes some time to render. Also, I'm planning to do Highlights & Annotations on the ebook where we need the book that the user is highlighting locally.

The workaround I'm planning is to keep the currently reading books locally. In other words, if you open a book for the first time, it will store it locally and the next time you open, it will fetch from the Joyread instance instead from Nextcloud. If the book is updated in any form - currently reading page, highlights, annotations - it will update the book in the Nextcloud.

And we can keep a limit for that - like how many books you want to keep it locally in the Joyread instance. Or we can set a limit for currently reading books (which is 10 books for each user at the moment).

This is a UX problem, I'll try to do this and come back. Meanwhile, if anybody has any better solution, Please let me know. Thanks

**Comment by GAumala** [comment]
Caching the books on the server seems good to me.
